### ISSUE TITLE
Panic index error when trying to upgrade node osDiskSizeGB

### ISSUE THREAD

**Issue by sblinz** [Azure/aks-engine#3437]
**Describe the bug**

I have an existing K8S cluster deployed with aks-engine that I want to upgrade the node osDiskSizeGB value. When I run aks-engine upgrade it crashes with a non-descriptive panic index error

**Steps To Reproduce**

Command executed: 

aks-engine upgrade -f  --subscription-id XXX     --client-id XXX     --client-secret XXX     --resource-group <my_rg>     --location <my_location>  --upgrade-version 1.17 --kubeconfig ~/.kube/config --api-model kubernetes-azure.json

API model (same as the one used to deploy except added `agentPoolProfiles.osDiskSizeGB: 128`.

{
  "apiVersion": "vlabs",
  "properties": {
    "orchestratorProfile": {
      "orchestratorType": "Kubernetes",
      "orchestratorRelease": "1.17",
      "kubernetesConfig": {
        "useManagedIdentity": false,
        "addons": [
          {
            "name": "blobfuse-flexvolume",
            "enabled": false
          },
          {
            "name": "keyvault-flexvolume",
            "enabled": false
          },
          {
            "name": "nvidia-device-plugin",
            "enabled": false
          },
          {
            "name": "tiller",
            "enabled": false
          },
          {
            "name": "kubernetes-dashboard",
            "enabled": false
          },
          {
            "name": "cluster-autoscaler",
            "enabled": true,
            "pools": [
              {
                "name": "agentpool",
                "config": {
                  "min-nodes": "0",
                  "max-nodes": "20"
                }
              }
            ],
            "config": {
              "scan-interval": "10s",
              "max-total-unready-percentage": "80",
              "skip-nodes-with-system-pods": "false",
              "unremovable-node-recheck-timeout": "10s",
              "scale-down-unneeded-time": "10s",
              "scale-down-delay-after-add": "10s",
              "scale-down-delay-after-failure": "10s",
              "scale-down-delay-after-delete": "10s"
            }
          }
        ]
      }
    },
    "masterProfile": {
      "count": 1,
      "dnsPrefix": "<VPC>",
      "vmSize": "Standard_D4_v3",
      "vnetSubnetId": "/subscriptions/<SUBSCRIPTION_ID>/resourceGroups/<VPC>-rg/providers/Microsoft.Network/virtualNetworks/<VPC>-vnet/subnets/<VPC>-k8s",
      "firstConsecutiveStaticIP": "10.x.32.250"
    },
    "agentPoolProfiles": [
      {
        "name": "agentpool",
        "count": 1,
        "vmSize": "Standard_NV6",
        "osDiskSizeGB": 128,
        "imageReference": {
          "name": "k8s-node",
          "resourceGroup": "common-rg",
          "subscriptionId": "XXX",
          "gallery": "images",
          "version": "1.0.1"
        },        
        "availabilityProfile": "VirtualMachineScaleSets",
        "storageProfile": "ManagedDisks",
        "vnetSubnetId": "/subscriptions/<SUBSCRIPTION_ID>/resourceGroups/<VPC>-rg/providers/Microsoft.Network/virtualNetworks/<VPC>-vnet/subnets/<VPC>-k8s"
      }
    ],
    "linuxProfile": {
      "adminUsername": "azureuser",
      "ssh": {
        "publicKeys": [
          {
            "keyData": ""
          }
        ]
      }
    },
    "servicePrincipalProfile": {
      "clientId": "XXX",
      "secret": "XXX"
    }
  }
}

**Expected behavior**

No error

**AKS Engine version**

0.48.0

**Kubernetes version**

1.17

**Additional context**

**Comment by sblinz** [comment]
@jackfrancis Thanks for the reply. What do you mean "full version"? The cluster was deployed with "1.17" and I don't want to change it, I only want to upgrade osDiskSizeGB.

**Comment by jackfrancis** [comment]
Which version of Kubernetes is it running (e.g., "1.17.6") You need to use the full version.

**Comment by sblinz** [comment]
Gotcha, thanks. I'm getting another error now and I wasn't sure if I should open another issue for this so I'll start by asking you here. If I should open a separate issue just let me know.

The upgrade goes all the way through deleting the master and finishing the ARM deployment with "Succeeded" output. But after that I get a ton of these messages:

**Comment by jackfrancis** [comment]
That error looks like the k8s certificate the apiserver is using is incorrect.

Do these values looks like what you expect in the api model you passed to `aks-engine upgrade`?

```
    "certificateProfile": {
      "caCertificate": "xxxx",
      "caPrivateKey": "xxxx",
      "apiServerCertificate": "xxxx",
      "apiServerPrivateKey": xxxx,
      "clientCertificate": "xxxx",
      "clientPrivateKey": "xxxx",
      "kubeConfigCertificate": "xxxx",
      "kubeConfigPrivateKey": xxxx,
      "etcdServerCertificate": "xxxx",
      "etcdServerPrivateKey": "xxxx",
      "etcdClientCertificate": "xxxx",
      "etcdClientPrivateKey": "xxxx",
      "etcdPeerCertificates": [
        "xxxx"
      ],
      "etcdPeerPrivateKeys": [
        "xxxx"
      ]
    },
```

**Comment by sblinz** [comment]
I used the same api model for the upgrade that we used when we created the cluster with aks-engine deploy. except that I added the osDiskSizeGB value to it. That api model file is included in the OP in its entirety and we did not have any certificateProfile object in it. We only provided linuxProfile > ssh > publicKeys > keyData. 

Based on what you wrote, I suspect I should have ran upgrade with the api model autogenerated by the deployment (the one that gets written to _output/***/apimodel.json). This one does have the certificateProfile object you showed with all the values filled out.

I tried re-running aks-engine upgrade using the _output api model (after adding the osDiskSizeGB value I wanted) but now I am getting the following error right away - my guess is I have to deploy the cluster from scratch because the upgrade I ran with the our starting api model file (the "incomplete" one above) broke it (for example, there is no master VM anymore).

INFO[0011] Retry updating cluster-autoscaler after 0 seconds
WARN[0011] Failed to update cluster-autoscaler deployment: Get https://snw.southeastasia.cloudapp.azure.com/apis/apps/v1/namespaces/kube-system/deployments/cluster-autoscaler: x509: certificate signed by unknown authority (possibly because of "crypto/rsa: verification error" while trying to verify candidate authority certificate "ca")

**Comment by sblinz** [comment]
Sorry, correction - I do have a master VM up and running, I missed it in my dashboard.

**Comment by jackfrancis** [comment]
@sblinz, correct, the api model JSON to use for all `aks-engine scale` and `aks-engine upgrade` operations is the fully configured/generated file after `aks-engine deploy` (or `aks-engine generate`). There's a fighting chance that if you re-run `aks-engine upgrade` against the resource group using the correct api model input you should be O.K. But, generally, if you *can* throw away that cluster it's probably easiest to build a new one.
